govcd/api_vcd.go

```diff
@@ -19,38 +19,30 @@ import (
 type VCDClientOption func(*VCDClient) error
 
 type VCDClient struct {
-	Client      Client  // Client for the underlying VCD instance
-	sessionHREF url.URL // HREF for the session API
-	QueryHREF   url.URL // HREF for the query API
-	Mutex       sync.Mutex
-}
-
-type supportedVersions struct {
-	VersionInfo struct {
-		Version  string `xml:"Version"`
-		LoginUrl string `xml:"LoginUrl"`
-	} `xml:"VersionInfo"`
+	Client            Client  // Client for the underlying VCD instance
+	sessionHREF       url.URL // HREF for the session API
+	QueryHREF         url.URL // HREF for the query API
+	Mutex             sync.Mutex
+	supportedVersions SupportedVersions // Versions from /api/versions endpoint
 }
 
 func (vdcCli *VCDClient) vcdloginurl() error {
-	apiEndpoint := vdcCli.Client.VCDHREF
-	apiEndpoint.Path += "/versions"
-	// No point in checking for errors here
-	req := vdcCli.Client.NewRequest(map[string]string{}, "GET", apiEndpoint, nil)
-	resp, err := checkResp(vdcCli.Client.Http.Do(req))
-	if err != nil {
-		return err
+	if err := vdcCli.validateAPIVersion(); err != nil {
+		return fmt.Errorf("could not find valid version for login: %s", err)
 	}
-	defer resp.Body.Close()
 
-	supportedVersions := new(supportedVersions)
-	err = decodeBody(resp, supportedVersions)
-	if err != nil {
-		return fmt.Errorf("error decoding versions response: %s", err)
+	// find login address matching the API version
+	var neededVersion VersionInfo
+	for _, v := range vdcCli.supportedVersions.VersionInfos {
```

Comment:
Let's try to avoid one letter variables

Reply:
I can change these, but that's idiomatic Go for limited scope vars - https://github.com/golang/go/wiki/CodeReviewComments#variable-names

Reply:
We prefer to avoid them: when we inherited the library, we had to understand what the variables in the loops meant. We saw the same `v` used to refer to a vcd, a vdc, a version, a vapp, a vm, and it took us some time to bring the algorithms to the light. 
It's better to invest a few letters to make the code readable without effort.

Reply:
2 vs 1 - agreed 👍

Reply:
Oh, Giuseppe was faster ;-)